## Security review — REQUEST CHANGES

Opus-tier review (package:security, supply-chain surface). The Strategy-C structure is right and the manifest is correctly url:null with no binaries committed, but there are blockers: CI is red on this PR's own files, and the verify gate has two path/verification defects that defeat the core acceptance criterion ("fails the build if any model is hash-mismatched").

### Blockers

**B1 — CI red: banned founder-identity literals in shipped/tracked files.**
The `Banlist + unit-name block` gate fails on this PR's own files:
- `third_party/models/manifest.yml:33` — `arlowe-1`
- `third_party/whisplay-driver/INSTALL.md:47,53,56,125,131` — `arlowe-1`, `arlowe-1.local`, `/home/focal55`

A red PR cannot be approved. Fix by sanitizing the literals (preferred — generalize to "the dev unit" / `$ARLOWE_WHISPLAY_SRC` / `<your-home>`), or, if a literal is genuinely load-bearing provenance, add the path to `.sanitize-allowlist` with rationale (precedent: `third_party/whisplay-driver/PROVENANCE.md`, `scripts/dev-pull-from-pi.sh`). Note the allowlist's own guidance: "does this file ship in the firmware image? If yes, fix the literal." Neither of these files ships, so allowlisting is defensible — but the manifest.yml `arlowe-1` reference reads as incidental and is cleaner to just remove.

**B2 — Real (non-placeholder) SHA pin on a DIRECTORY model silently degrades to a non-blocking WARN.**
The gate only recognizes `model.bin` as the "primary file" inside a directory artifact. For Qwen (`install_to: /opt/arlowe/models/qwen2.5-7b-int4-ax650`, a directory with no `model.bin`), even a real pin produces:
`[WARN] qwen2.5-7b-int4-ax650  present but no primary file found to verify sha256`
and the script can still exit 0. Reproduced locally: set the qwen pin to a real hash, stage the dir → WARN, not FAIL. This means the Qwen model — the artifact `qwen-api.service` actually loads via `QWEN_MODEL_DIR` — is **never integrity-verified** by this gate, regardless of pin. A tampered Qwen directory passes. That directly defeats the acceptance criterion and the whole point of the supply-chain gate. A real pin that cannot be verified must FAIL (set `all_ok=false`), not WARN. Decide on a concrete verifiable artifact per model family (e.g. hash a specific file, or a deterministic `find -type f | sort | xargs sha256sum | sha256sum` over the dir) and make "real pin present but unverifiable" a hard fail.

**B3 — Three inconsistent path schemes: manifest `install_to`, INSTALL.md placement, and the gate's actual search path do not agree.**
The gate derives its search name from `model.name`, so it looks for:
- `<DIR>/faster-whisper-small.en/model.bin`
- `<DIR>/piper-en_US-lessac-medium`

But INSTALL.md instructs builders to place artifacts at:
- `<DIR>/whisper/small.en/model.bin`
- `<DIR>/piper-voices/en_US-lessac-medium.onnx`

Reproduced: a piper artifact staged exactly per INSTALL.md is reported `[FAIL] not found`; it only passes when placed at the gate's literal `model.name` path. So a builder who follows the docs correctly gets a false failure, and the only way to make the gate pass is to stage the file somewhere that does NOT match the `install_to` the runtime reads. A gate that verifies a path the runtime never loads (and that contradicts its own install docs) is worse than no gate — it manufactures false confidence. Reconcile all three: the gate should locate the artifact at (or derived from) the manifest's `install_to`, and INSTALL.md must document that same path.

### Suggestions (not blocking)

- **S1 — ADR-0006 upstream link is a non-canonical fork.** It links faster-whisper to `github.com/guillaumekleinhans/faster-whisper` and lists `guillaumekleinhans/faster-whisper-*` as a model source. Canonical is `github.com/SYSTRAN/faster-whisper` / HF org `Systran`. In a supply-chain ADR a wrong provenance pointer could lead a future builder to fetch from the wrong place. Point at the canonical repo; the manifest already correctly uses `Systran/faster-whisper-small.en`.
- **S2 — Positional `read` parser is brittle.** `model_keys` is space-joined in Python and split with `IFS=' ' read -r key name sha install`. Today no field contains spaces, but if any `install_to`/`name` later does, fields misalign silently. Consider a tab or null delimiter.

### What I verified passes (no regression)

- `bash -n scripts/verify-third-party.sh` — clean.
- Existing axcl `.deb` SHA check and ax-llm submodule-commit check — intact, still fail correctly when absent.
- Missing-artifact path → hard FAIL, exit 1 (correct).
- Placeholder pin + artifact present → WARN, prints actual hash, does NOT set `all_ok=false` (correct — and there is no path where a placeholder is reported as `[OK]`/"passed").
- Real pin + correct hash on a `model.bin`-bearing dir → `[OK]`; real pin + wrong hash → `[FAIL]` exit 1 (correct for that shape).
- Strategy-C compliance: `url: null`, no model/driver binaries committed on the branch.
- No injection surface: filenames come from the in-repo manifest (trusted), `sha256sum`/`printf` args are quoted, no `eval`/dynamic execution.

shellcheck could not run locally (Mac, no shellcheck — matches the PR note); the pre-existing `printf "${COLOR}..."` (SC2059) pattern is mirrored consistently and was green on main, so not flagged.

Net: B2 and B3 together mean the gate, as written, would let a corrupt/wrong Qwen model through and false-fail a correctly-staged Whisper/Piper model. Fix the three blockers and re-request review.

--- a/docs/architecture/0006-whisper-model-selection.md
+++ b/docs/architecture/0006-whisper-model-selection.md
@@ -8,7 +8,7 @@
 
 ## Context
 
-The `whisper-stt.service` uses [faster-whisper](https://github.com/guillaumekleinhans/faster-whisper)
+The `whisper-stt.service` uses [faster-whisper](https://github.com/SYSTRAN/faster-whisper)
 to run on-device speech-to-text. The model must be pinned to a single concrete
 choice before the Phase 6 image build can populate the shared models partition.
 
@@ -73,4 +73,4 @@ to `small.en`) is the mechanism for per-device overrides without reflashing.
 - Shared models partition decision: `docs/architecture/0004-shared-models-partition.md` (ADR-0004)
 - Model manifest: `third_party/models/manifest.yml`
 - STT server: `runtime/stt/stt_server.py`
-- HuggingFace faster-whisper models: `guillaumekleinhans/faster-whisper-*` or `Systran/faster-whisper-small.en`
+- HuggingFace faster-whisper models: `Systran/faster-whisper-small.en` (canonical SYSTRAN org)

--- a/scripts/verify-third-party.sh
+++ b/scripts/verify-third-party.sh
@@ -20,6 +20,10 @@ MODELS_MANIFEST="${REPO_ROOT}/third_party/models/manifest.yml"
 AX_LLM_DIR="${REPO_ROOT}/third_party/ax-llm"
 PINNED_AXLLM_COMMIT="df75c34ca2ed8fe55e7576204e4da9c5b5f88ad8"
 
+# install_to paths in the manifest are image-absolute: /opt/arlowe/models/<subpath>.
+# The gate strips this prefix to get the subpath, then searches under the configured roots.
+MODELS_IMAGE_PREFIX="/opt/arlowe/models"
+
 RED='\033[0;31m'
 GREEN='\033[0;32m'
 YELLOW='\033[1;33m'
@@ -37,10 +41,15 @@ Checks:
   3. Model artifacts (Qwen LLM, Whisper STT, Piper TTS) per third_party/models/manifest.yml
   4. WhisPlay driver source (WhisPlay.py + LICENSE) is locatable
 
-Model artifact search order (per artifact):
-  - \$ARLOWE_MODELS_DIR/<artifact-name>/
-  - third_party/models/<artifact-name>/
-  - /var/cache/arlowe-build/models/<artifact-name>/
+Model artifact search order (per artifact, using install_to subpath from manifest):
+  - \$ARLOWE_MODELS_DIR/<install_to-subpath>
+  - third_party/models/<install_to-subpath>
+  - /var/cache/arlowe-build/models/<install_to-subpath>
+
+The install_to subpath is derived by stripping the image-side prefix
+(${MODELS_IMAGE_PREFIX}) from each manifest's install_to field.
+This ensures the gate locates artifacts at the same relative path the
+runtime units read — no separate "search name" that can drift from install_to.
 
 WhisPlay driver search order:
   - \$ARLOWE_WHISPLAY_SRC/WhisPlay.py
@@ -63,6 +72,16 @@ if [[ "${1:-}" == "--help" || "${1:-}" == "-h" ]]; then
 fi
 
 all_ok=true
+
+# ---------------------------------------------------------------------------
+# Helper: compute a deterministic digest over a directory.
+# Walks all regular files under $1 in sorted order, hashes each, then hashes
+# the combined output. Reproducible across runs on the same tree.
+# ---------------------------------------------------------------------------
+dir_sha256() {
+  local dir="$1"
+  find "${dir}" -type f | LC_ALL=C sort | xargs sha256sum | sha256sum | awk '{print $1}'
+}
 
 # ---------------------------------------------------------------------------
 # Read expected SHA-256 from manifest.yml
@@ -146,81 +165,78 @@ fi
 
 # ---------------------------------------------------------------------------
 # Check 3: Model artifacts (Qwen LLM, Whisper STT, Piper TTS)
+#
+# Path scheme (unified — B3 fix):
+#   Each manifest entry has an install_to field with the image-absolute path
+#   (e.g. /opt/arlowe/models/whisper/small.en). The gate strips MODELS_IMAGE_PREFIX
+#   to derive the relative subpath (e.g. whisper/small.en), then searches:
+#     $ARLOWE_MODELS_DIR/<subpath>
+#     third_party/models/<subpath>
+#     /var/cache/arlowe-build/models/<subpath>
+#   This matches INSTALL.md staging instructions exactly.
+#
+# Directory artifacts with real SHA pins (B2 fix):
+#   A real (non-placeholder) pin on a directory artifact is verified by computing
+#   a deterministic directory digest (sorted find -type f | xargs sha256sum | sha256sum).
+#   A real pin with no verifiable target (missing primary_file and not a verifiable
+#   directory) HARD-FAILS — WARN is only for TODO placeholders.
 # ---------------------------------------------------------------------------
 if [[ ! -f "${MODELS_MANIFEST}" ]]; then
   printf "${RED}[FAIL]${NC} third_party/models/manifest.yml  not found\n"
   echo >&2 "  Expected at: ${MODELS_MANIFEST}"
   all_ok=false
 else
-  # Each model entry: name, sha256, install_to
-  # Read via python3 (same dependency as axcl check above — no new dep).
-  model_count=$(python3 -c "
-import yaml
-with open('${MODELS_MANIFEST}') as f:
-    m = yaml.safe_load(f)
-print(len(m.get('models', {})))
-" 2>/dev/null) || {
-    echo >&2 "ERROR: failed to parse ${MODELS_MANIFEST} (is python3-yaml installed?)"
-    exit 1
-  }
-
+  # Emit tab-delimited fields to survive spaces in any future install_to values (S2 fix).
   model_keys=$(python3 -c "
 import yaml
 with open('${MODELS_MANIFEST}') as f:
     m = yaml.safe_load(f)
 for key in m.get('models', {}):
     entry = m['models'][key]
-    print(key, entry.get('name',''), entry.get('sha256',''), entry.get('install_to',''))
-" 2>/dev/null)
+    fields = [key, entry.get('name',''), entry.get('sha256',''), entry.get('install_to','')]
+    print('\t'.join(fields))
+" 2>/dev/null) || {
+    echo >&2 "ERROR: failed to parse ${MODELS_MANIFEST} (is python3-yaml installed?)"
+    exit 1
+  }
 
-  while IFS=' ' read -r model_key model_name model_sha model_install; do
+  while IFS=$'\t' read -r model_key model_name model_sha model_install; do
     [[ -z "${model_key}" ]] && continue
 
-    # Locate the artifact directory or primary file.
-    # For directory-based models (qwen, whisper CTranslate2): check for the dir.
-    # For file-based models (piper .onnx): check for the file.
+    # Derive the relative subpath from install_to by stripping the image prefix.
+    # install_to: /opt/arlowe/models/whisper/small.en  →  subpath: whisper/small.en
+    if [[ "${model_install}" == "${MODELS_IMAGE_PREFIX}"/* ]]; then
+      install_subpath="${model_install#${MODELS_IMAGE_PREFIX}/}"
+    else
+      # install_to does not start with the expected prefix — fall back to model name.
+      install_subpath="${model_name}"
+    fi
+
+    # Search for the artifact using the install_to subpath.
     artifact_path=""
 
-    # Derive a search name from the artifact name (strip version suffixes for dir check)
-    search_name="${model_name}"
-
     if [[ -n "${ARLOWE_MODELS_DIR:-}" ]]; then
-      candidate="${ARLOWE_MODELS_DIR}/${search_name}"
+      candidate="${ARLOWE_MODELS_DIR}/${install_subpath}"
       if [[ -e "${candidate}" ]]; then
         artifact_path="${candidate}"
       fi
     fi
 
     if [[ -z "${artifact_path}" ]]; then
-      candidate="${REPO_ROOT}/third_party/models/${search_name}"
+      candidate="${REPO_ROOT}/third_party/models/${install_subpath}"
       if [[ -e "${candidate}" ]]; then
         artifact_path="${candidate}"
       fi
     fi
 
     if [[ -z "${artifact_path}" ]]; then
-      candidate="/var/cache/arlowe-build/models/${search_name}"
+      candidate="/var/cache/arlowe-build/models/${install_subpath}"
       if [[ -e "${candidate}" ]]; then
         artifact_path="${candidate}"
       fi
     fi
 
-    # Determine the primary verifiable file inside the artifact location.
-    # Whisper: model.bin; Piper: the .onnx file itself (install_to path); Qwen: any single large file.
-    primary_file=""
-    if [[ -n "${artifact_path}" ]]; then
-      if [[ -f "${artifact_path}" ]]; then
-        # Direct file (piper .onnx path)
-        primary_file="${artifact_path}"
-      elif [[ -d "${artifact_path}" ]]; then
-        # Directory — look for model.bin (faster-whisper CTranslate2)
-        if [[ -f "${artifact_path}/model.bin" ]]; then
-          primary_file="${artifact_path}/model.bin"
-        fi
-      fi
-    fi
-
-    # Check if the SHA pin is a TODO placeholder
+    # Check if the SHA pin is a TODO placeholder.
     is_placeholder=false
     if [[ "${model_sha}" == TODO_SHA256* ]]; then
       is_placeholder=true
@@ -228,26 +244,32 @@ for key in m.get('models', {}):
 
     if [[ -z "${artifact_path}" ]]; then
       printf "${RED}[FAIL]${NC} %-50s not found\n" "${model_name}"
-      echo >&2 "  Set ARLOWE_MODELS_DIR=<dir> or place artifact at:"
-      echo >&2 "    third_party/models/${search_name}/"
-      echo >&2 "    /var/cache/arlowe-build/models/${search_name}/"
+      echo >&2 "  Set ARLOWE_MODELS_DIR=<dir> or stage artifact at:"
+      echo >&2 "    third_party/models/${install_subpath}"
+      echo >&2 "    /var/cache/arlowe-build/models/${install_subpath}"
       echo >&2 "  See third_party/models/INSTALL.md for sourcing instructions."
       all_ok=false
+
     elif [[ "${is_placeholder}" == "true" ]]; then
-      # Placeholder pin — warn, print actual hash if we can compute it
-      if [[ -n "${primary_file}" ]]; then
-        actual_hash=$(sha256sum "${primary_file}" | awk '{print $1}')
+      # Placeholder pin — warn and print actual hash if computable, never fail.
+      if [[ -f "${artifact_path}" ]]; then
+        actual_hash=$(sha256sum "${artifact_path}" | awk '{print $1}')
         printf "${YELLOW}[WARN]${NC}  %-50s sha256 pin is TODO placeholder\n" "${model_name}"
-        printf "         actual hash of %s:\n" "$(basename "${primary_file}")"
-        printf "         %s\n" "${actual_hash}"
+        printf "         actual hash: %s\n" "${actual_hash}"
+        printf "         Record this in third_party/models/manifest.yml to close the TODO.\n"
+      elif [[ -d "${artifact_path}" ]]; then
+        actual_hash=$(dir_sha256 "${artifact_path}")
+        printf "${YELLOW}[WARN]${NC}  %-50s sha256 pin is TODO placeholder\n" "${model_name}"
+        printf "         actual dir digest: %s\n" "${actual_hash}"
         printf "         Record this in third_party/models/manifest.yml to close the TODO.\n"
       else
-        printf "${YELLOW}[WARN]${NC}  %-50s sha256 pin is TODO placeholder; artifact present but no primary file found to hash\n" "${model_name}"
+        printf "${YELLOW}[WARN]${NC}  %-50s sha256 pin is TODO placeholder; artifact present but unreadable\n" "${model_name}"
       fi
+
     else
-      # Real SHA pin — verify
-      if [[ -n "${primary_file}" ]]; then
-        actual_sha256=$(sha256sum "${primary_file}" | awk '{print $1}')
+      # Real SHA pin — must verify; no degradation to WARN.
+      if [[ -f "${artifact_path}" ]]; then
+        actual_sha256=$(sha256sum "${artifact_path}" | awk '{print $1}')
         if [[ "${actual_sha256}" == "${model_sha}" ]]; then
           printf "${GREEN}[OK]${NC}   %-50s sha256 matches\n" "${model_name}"
         else
@@ -256,8 +278,23 @@ for key in m.get('models', {}):
           echo >&2 "  Actual:   ${actual_sha256}"
           all_ok=false
         fi
+      elif [[ -d "${artifact_path}" ]]; then
+        actual_sha256=$(dir_sha256 "${artifact_path}")
+        if [[ "${actual_sha256}" == "${model_sha}" ]]; then
+          printf "${GREEN}[OK]${NC}   %-50s dir digest matches\n" "${model_name}"
+        else
+          printf "${RED}[FAIL]${NC} %-50s dir digest mismatch\n" "${model_name}"
+          echo >&2 "  Expected: ${model_sha}"
+          echo >&2 "  Actual:   ${actual_sha256}"
+          echo >&2 "  Digest is sha256(sorted find -type f | xargs sha256sum | sha256sum)"
+          all_ok=false
+        fi
       else
-        printf "${YELLOW}[WARN]${NC}  %-50s present but no primary file found to verify sha256\n" "${model_name}"
+        # Present as a path but neither file nor directory — treat as FAIL,
+        # same as a real pin with no verifiable target.
+        printf "${RED}[FAIL]${NC} %-50s real sha256 pin present but artifact is not a file or directory\n" "${model_name}"
+        echo >&2 "  Cannot verify ${artifact_path} — check staging."
+        all_ok=false
       fi
     fi
   done <<< "${model_keys}"

